<a href="https://cognitiveclass.ai/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkRP0321ENSkillsNetwork25371262-2022-01-01">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-RP0101EN-Coursera/v2/M1_R_Basics/images/IDSNlogo.png" width="200" align="center">
</a>


<h1>Data Wrangling with dplyr</h1>

Estimated time needed: **60** minutes


## Lab Overview:

In this lab, you will focus on wrangling the Seoul bike-sharing demand historical dataset. This is the core dataset to build a predictive model later.

It contains the following columns:

*   `DATE` : Year-month-day
*   `RENTED BIKE COUNT`- Count of bikes rented at each hour
*   `HOUR`- Hour of he day
*   `TEMPERATURE` - Temperature in Celsius
*   `HUMIDITY` - Unit is `%`
*   `WINDSPEED` - Unit is `m/s`
*   `VISIBILITY` - Multiplied by 10m
*   `DEW POINT TEMERATURE` - The temperature to which the air would have to cool down in order to reach saturation, unit is Celsius
*   `SOLAR RADIATION` - MJ/m2
*   `RAINFALL` - mm
*   `SNOWFALL` - cm
*   `SEASONS` - Winter, Spring, Summer, Autumn
*   `HOLIDAY` - Holiday/No holiday
*   `FUNCTIONAL DAY` - NoFunc(Non Functional Hours), Fun(Functional hours)

For this dataset, you will be asked to use `tidyverse` to perform the following data wrangling tasks:

*   **TASK: Detect and handle missing values**
*   **TASK: Create indicator (dummy) variables for categorical variables**
*   **TASK: Normalize data**

Let's start!


First import the necessary library for this data wrangling task:


In [1]:
# Check if you need to install the `tidyverse` library
require("tidyverse")
library(tidyverse)

Loading required package: tidyverse
── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──
✔ ggplot2 3.3.0     ✔ purrr   0.3.4
✔ tibble  3.0.1     ✔ dplyr   0.8.5
✔ tidyr   1.0.2     ✔ stringr 1.4.0
✔ readr   1.3.1     ✔ forcats 0.5.0
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()


Then load the bike-sharing system data from the csv processed in the previous lab:


In [2]:
bike_sharing_df <- read_csv("raw_seoul_bike_sharing.csv")

Parsed with column specification:
cols(
  DATE = col_character(),
  RENTED_BIKE_COUNT = col_double(),
  HOUR = col_double(),
  TEMPERATURE = col_double(),
  HUMIDITY = col_double(),
  WIND_SPEED = col_double(),
  VISIBILITY = col_double(),
  DEW_POINT_TEMPERATURE = col_double(),
  SOLAR_RADIATION = col_double(),
  RAINFALL = col_double(),
  SNOWFALL = col_double(),
  SEASONS = col_character(),
  HOLIDAY = col_character(),
  FUNCTIONING_DAY = col_character()
)


In [ ]:
# Or you may read it from here again
# url <- "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-RP0321EN-SkillsNetwork/labs/datasets/raw_seoul_bike_sharing.csv"
# Notice some column names in the raw datasets are not standalized if you haven't done them properly in the previous lab

First take a quick look at the dataset:


In [3]:
summary(bike_sharing_df)
dim(bike_sharing_df)

     DATE           RENTED_BIKE_COUNT      HOUR        TEMPERATURE    
 Length:8760        Min.   :   2.0    Min.   : 0.00   Min.   :-17.80  
 Class :character   1st Qu.: 214.0    1st Qu.: 5.75   1st Qu.:  3.40  
 Mode  :character   Median : 542.0    Median :11.50   Median : 13.70  
                    Mean   : 729.2    Mean   :11.50   Mean   : 12.87  
                    3rd Qu.:1084.0    3rd Qu.:17.25   3rd Qu.: 22.50  
                    Max.   :3556.0    Max.   :23.00   Max.   : 39.40  
                    NA's   :295                       NA's   :11      
    HUMIDITY       WIND_SPEED      VISIBILITY   DEW_POINT_TEMPERATURE
 Min.   : 0.00   Min.   :0.000   Min.   :  27   Min.   :-30.600      
 1st Qu.:42.00   1st Qu.:0.900   1st Qu.: 940   1st Qu.: -4.700      
 Median :57.00   Median :1.500   Median :1698   Median :  5.100      
 Mean   :58.23   Mean   :1.725   Mean   :1437   Mean   :  4.074      
 3rd Qu.:74.00   3rd Qu.:2.300   3rd Qu.:2000   3rd Qu.: 14.800      
 Max.   :98.

[1] 8760   14

From the summary, we can observe that:

Columns `RENTED_BIKE_COUNT`, `TEMPERATURE`, `HUMIDITY`, `WIND_SPEED`, `VISIBILITY`, `DEW_POINT_TEMPERATURE`, `SOLAR_RADIATION`, `RAINFALL`, `SNOWFALL` are numerical variables/columns and require normalization. Moreover, `RENTED_BIKE_COUNT` and `TEMPERATURE` have some missing values (NA's) that need to be handled properly.

`SEASONS`, `HOLIDAY`, `FUNCTIONING_DAY` are categorical variables which need to be converted into indicator columns or dummy variables.
Also, `HOUR` is read as a numerical variable but it is in fact a categorical variable with levels ranging from 0 to 23.

Now that you have some basic ideas about how to process this bike-sharing demand dataset, let's start working on it!


# TASK: Detect and handle missing values


The `RENTED_BIKE_COUNT` column has about 295 missing values, and `TEMPERATURE` has about 11 missing values. Those missing values could be caused by not being recorded, or from malfunctioning bike-sharing systems or weather sensor networks. In any cases, the identified missing values have to be properly handled.


Let's first handle missing values in `RENTED_BIKE_COUNT` column:


Considering `RENTED_BIKE_COUNT` is the response variable/dependent variable, i.e., we want to predict the `RENTED_BIKE_COUNT` using other predictor/independent variables later, and we normally can not allow missing values for the response variable, so missing values for response variable must be either dropped or imputed properly.

We can see that `RENTED_BIKE_COUNT` only has about 3% missing values (295 / 8760). As such, you can safely drop any rows whose `RENTED_BIKE_COUNT` has missing values.


*TODO:* Drop rows with missing values in the `RENTED_BIKE_COUNT` column


In [4]:
# Drop rows with `RENTED_BIKE_COUNT` column == NA
bike_sharing_df %>% drop_na(RENTED_BIKE_COUNT)

DATE       RENTED_BIKE_COUNT HOUR TEMPERATURE HUMIDITY WIND_SPEED
1    01/12/2017 254                0   -5.2        37       2.2       
2    01/12/2017 204                1   -5.5        38       0.8       
3    01/12/2017 173                2   -6.0        39       1.0       
4    01/12/2017 107                3   -6.2        40       0.9       
5    01/12/2017  78                4   -6.0        36       2.3       
6    01/12/2017 100                5   -6.4        37       1.5       
7    01/12/2017 181                6   -6.6        35       1.3       
8    01/12/2017 460                7   -7.4        38       0.9       
9    01/12/2017 930                8   -7.6        37       1.1       
10   01/12/2017 490                9   -6.5        27       0.5       
11   01/12/2017 339               10   -3.5        24       1.2       
12   01/12/2017 360               11   -0.5        21       1.3       
13   01/12/2017 449               12    1.7        23       1.4       
14   01/12/2017 451               13    2.4        25       1.6       
15   01/12/2017 447               14    3.0        26       2.0       
16   01/12/2017 463               15    2.1        36       3.2       
17   01/12/2017 484               16    1.2        54       4.2       
18   01/12/2017 555               17    0.8        58       1.6       
19   01/12/2017 862               18    0.6        66       1.4       
20   01/12/2017 600               19    0.0        77       1.7       
21   01/12/2017 426               20   -0.3        79       1.5       
22   01/12/2017 405               21   -0.8        81       0.8       
23   01/12/2017 398               22   -0.9        83       1.5       
24   01/12/2017 323               23   -1.3        84       1.0       
25   02/12/2017 328                0   -1.8        87       1.1       
26   02/12/2017 308                1   -2.2        86       0.6       
27   02/12/2017 262                2   -2.9        86       1.5       
28   02/12/2017 167                3   -3.5        81       2.2       
29   02/12/2017  89                4   -3.8        79       2.0       
30   02/12/2017  79                5   -4.0        79       1.5       
⋮    ⋮          ⋮                 ⋮    ⋮           ⋮        ⋮         
8436 29/11/2018 1577              18    0.0        46       0.9       
8437 29/11/2018 1026              19    5.2        48       1.9       
8438 29/11/2018  818              20    4.1        54       1.3       
8439 29/11/2018  792              21    3.8        55       2.1       
8440 29/11/2018  726              22    3.0        59       1.1       
8441 29/11/2018  559              23    2.6        56       1.9       
8442 30/11/2018  474               0    1.8        62       1.6       
8443 30/11/2018  456               1    1.3        63       2.0       
8444 30/11/2018  300               2    0.6        66       0.5       
8445 30/11/2018  213               3   -0.1        69       0.9       
8446 30/11/2018  116               4   -0.5        71       0.4       
8447 30/11/2018  149               5   -0.7        66       0.5       
8448 30/11/2018  293               6   -0.8        68       0.8       
8449 30/11/2018  750               7   -1.2        70       0.8       
8450 30/11/2018 1527               8   -1.5        68       1.1       
8451 30/11/2018  809               9   -0.4        57       0.6       
8452 30/11/2018  554              10    1.9        51       0.8       
8453 30/11/2018  642              11    5.3        43       1.8       
8454 30/11/2018  720              12    6.6        35       1.3       
8455 30/11/2018  740              13    7.1        24       2.8       
8456 30/11/2018  761              14    7.8        20       2.2       
8457 30/11/2018  768              15    7.0        20       3.3       
8458 30/11/2018  837              16    7.2        23       1.5       
8459 30/11/2018 1047              17    6.0        29       2.1       
8460 30/11/2018 1384         

In [6]:
# Print the dataset dimension again after those rows are dropped
dim(bike_sharing_df)

[1] 8760   14

Now that you have handled  missing values in the `RENTED_BIKE_COUNT` variable, let's continue processing missing values for the `TEMPERATURE` column.


Unlike the `RENTED_BIKE_COUNT` variable, `TEMPERATURE` is not a response variable. However, it is still an important predictor variable - as you could imagine, there may be a positve correlation between `TEMPERATURE` and `RENTED_BIKE_COUNT`. For example, in winter time with lower temperatures, people may not want to ride a bike, while in summer with nicer weather, they are more likely to rent a bike.


How do we handle missing values for `TEMPERATURE`? We could simply remove the rows but it's better to impute them because `TEMPERATURE` should be relatively easy and reliable to estimate statistically.


Let's first take a look at the missing values in the TEMPERATURE column.


In [7]:
bike_sharing_df %>% 
                filter(is.na(TEMPERATURE))

DATE       RENTED_BIKE_COUNT HOUR TEMPERATURE HUMIDITY WIND_SPEED VISIBILITY
1  07/06/2018 3221              18   NA          57       2.7        1217      
2  12/06/2018 1246              14   NA          45       2.2        1961      
3  13/06/2018 2664              17   NA          57       3.3         919      
4  17/06/2018 2330              17   NA          58       3.3         865      
5  20/06/2018 2741              19   NA          61       2.7        1236      
6  30/06/2018 1144              13   NA          87       1.7         390      
7  05/07/2018  827              10   NA          75       1.1        1028      
8  11/07/2018  634               9   NA          96       0.6         450      
9  12/07/2018  593               6   NA          93       1.1         852      
10 21/07/2018  347               4   NA          77       1.2        1203      
11 21/08/2018 1277              23   NA          75       0.1        1892      
   DEW_POINT_TEMPERATURE SOLAR_RADIATION RAINFALL SNOWFALL SEASONS HOLIDAY   
1  16.4                  0.96            0.0      0        Summer  No Holiday
2  12.7                  1.39            0.0      0        Summer  No Holiday
3  16.4                  0.87            0.0      0        Summer  No Holiday
4  16.7                  0.66            0.0      0        Summer  No Holiday
5  17.5                  0.60            0.0      0        Summer  No Holiday
6  23.2                  0.71            3.5      0        Summer  No Holiday
7  20.8                  1.22            0.0      0        Summer  No Holiday
8  24.9                  0.41            0.0      0        Summer  No Holiday
9  24.3                  0.01            0.0      0        Summer  No Holiday
10 21.2                  0.00            0.0      0        Summer  No Holiday
11 20.8                  0.00            0.0      0        Summer  No Holiday
   FUNCTIONING_DAY
1  Yes            
2  Yes            
3  Yes            
4  Yes            
5  Yes            
6  Yes            
7  Yes            
8  Yes            
9  Yes            
10 Yes            
11 Yes

It seems that all of the missing values for `TEMPERATURE` are found in `SEASONS == Summer`, so it is reasonable to impute those missing values with the summer average temperature.


*TODO:* Impute missing values for the TEMPERATURE column using its mean value.


In [17]:
# Calculate the summer average temperature
temp_data <- subset(bike_sharing_df, !is.na(TEMPERATURE)&SEASONS == "Summer")
summer_avg <- mean(temp_data)

Warning message in mean.default(temp_data):
“argument is not numeric or logical: returning NA”

In [21]:
# Impute missing values for TEMPERATURE column with summer average temperature
bike_sharing_df <- bike_sharing_df %>% mutate(TEMPERATURE = replace_na(TEMPERATURE, summer_avg))

In [22]:
# Print the summary of the dataset again to make sure no missing values in all columns
summary(bike_sharing_df)

     DATE           RENTED_BIKE_COUNT      HOUR        TEMPERATURE    
 Length:8760        Min.   :   2.0    Min.   : 0.00   Min.   :-17.80  
 Class :character   1st Qu.: 214.0    1st Qu.: 5.75   1st Qu.:  3.40  
 Mode  :character   Median : 542.0    Median :11.50   Median : 13.70  
                    Mean   : 729.2    Mean   :11.50   Mean   : 12.87  
                    3rd Qu.:1084.0    3rd Qu.:17.25   3rd Qu.: 22.50  
                    Max.   :3556.0    Max.   :23.00   Max.   : 39.40  
                    NA's   :295                       NA's   :11      
    HUMIDITY       WIND_SPEED      VISIBILITY   DEW_POINT_TEMPERATURE
 Min.   : 0.00   Min.   :0.000   Min.   :  27   Min.   :-30.600      
 1st Qu.:42.00   1st Qu.:0.900   1st Qu.: 940   1st Qu.: -4.700      
 Median :57.00   Median :1.500   Median :1698   Median :  5.100      
 Mean   :58.23   Mean   :1.725   Mean   :1437   Mean   :  4.074      
 3rd Qu.:74.00   3rd Qu.:2.300   3rd Qu.:2000   3rd Qu.: 14.800      
 Max.   :98.

In [23]:
# Save the dataset as `seoul_bike_sharing.csv`
write.csv(bike_sharing_df, "seoul_bike_sharing.csv")

# TASK: Create indicator (dummy) variables for categorical variables


Regression models can not process categorical variables directly, thus we need to convert them into indicator variables.


In the bike-sharing demand dataset, `SEASONS`, `HOLIDAY`, `FUNCTIONING_DAY` are categorical variables.
Also, `HOUR` is read as a numerical variable but it is in fact a categorical variable with levels ranged from 0 to 23.


*TODO:* Convert `HOUR` column from numeric into character first:


In [24]:
# Using mutate() function to convert HOUR column into character type
bike_sharing_df <- bike_sharing_df %>% mutate(HOUR = as.character(HOUR))

`SEASONS`, `HOLIDAY`, `FUNCTIONING_DAY`,  `HOUR` are all character columns now and are ready to be converted into indicator variables.

For example, `SEASONS` has four categorical values: `Spring`, `Summer`, `Autumn`, `Winter`. We thus need to create four indicator/dummy variables `Spring`, `Summer`, `Autumn`, and `Winter` which only have the value 0 or 1.

So, given a data entry with the value `Spring` in the `SEASONS` column, the values for the four new columns `Spring`, `Summer`, `Autumn`, and `Winter` will be set to 1 for `Spring` and 0 for the others:

| Spring | Summer | Autumn | Winter |
| ------ | ------ | ------ | ------ |
| 1      | 0      | 0      | 0      |


*TODO:* Convert `SEASONS`, `HOLIDAY`, `FUNCTIONING_DAY`, and `HOUR` columns into indicator columns.

Note that if `FUNCTIONING_DAY` only contains one categorical value after missing values removal, then you don't need to convert it to an indicator column.


In [32]:
# Convert SEASONS, HOLIDAY, FUNCTIONING_DAY, and HOUR columns into indicator columns.
dataframe <- bike_sharing_df %>% mutate(dummy=1, dummy2 = 1, dummy3 = 1, dummy4 = 1) %>% spread(key=SEASONS, value = dummy, fill = 0)%>% spread(key=HOLIDAY, dummy2, fill = 0) %>% spread(key=FUNCTIONING_DAY, value = dummy3, fill = 0) %>% spread(key=HOUR, value = dummy4, fill = 0)

In [34]:
# Print the dataset summary again to make sure the indicator columns are created properly
summary(dataframe)

     DATE           RENTED_BIKE_COUNT  TEMPERATURE        HUMIDITY    
 Length:8760        Min.   :   2.0    Min.   :-17.80   Min.   : 0.00  
 Class :character   1st Qu.: 214.0    1st Qu.:  3.40   1st Qu.:42.00  
 Mode  :character   Median : 542.0    Median : 13.70   Median :57.00  
                    Mean   : 729.2    Mean   : 12.87   Mean   :58.23  
                    3rd Qu.:1084.0    3rd Qu.: 22.50   3rd Qu.:74.00  
                    Max.   :3556.0    Max.   : 39.40   Max.   :98.00  
                    NA's   :295       NA's   :11                      
   WIND_SPEED      VISIBILITY   DEW_POINT_TEMPERATURE SOLAR_RADIATION 
 Min.   :0.000   Min.   :  27   Min.   :-30.600       Min.   :0.0000  
 1st Qu.:0.900   1st Qu.: 940   1st Qu.: -4.700       1st Qu.:0.0000  
 Median :1.500   Median :1698   Median :  5.100       Median :0.0100  
 Mean   :1.725   Mean   :1437   Mean   :  4.074       Mean   :0.5691  
 3rd Qu.:2.300   3rd Qu.:2000   3rd Qu.: 14.800       3rd Qu.:0.9300  
 Max. 

In [35]:
# Save the dataset as `seoul_bike_sharing_converted.csv`
write_csv(dataframe, "seoul_bike_sharing_converted.csv")

# TASK: Normalize data


Columns `RENTED_BIKE_COUNT`, `TEMPERATURE`, `HUMIDITY`, `WIND_SPEED`, `VISIBILITY`, `DEW_POINT_TEMPERATURE`, `SOLAR_RADIATION`, `RAINFALL`, `SNOWFALL` are numerical variables/columns with different value units and range. Columns with large values may adversely influence (bias) the predictive models and degrade model accuracy. Thus, we need to perform normalization on these numeric columns to transfer them into a similar range.


In this project, you are asked to use Min-max normalization:

**Min-max** rescales each value in a column by first subtracting the minimum value of the column from each value, and then divides the result by the difference between the maximum and minimum values of the column. So the column gets re-scaled such that the minimum becomes 0 and the maximum becomes 1.

$$x\_{new} = \frac{x\_{old} - x\_{min}}{x\_{max} - x\_{min}}$$


*TODO:* Apply min-max normalization on `RENTED_BIKE_COUNT`, `TEMPERATURE`, `HUMIDITY`, `WIND_SPEED`, `VISIBILITY`, `DEW_POINT_TEMPERATURE`, `SOLAR_RADIATION`, `RAINFALL`, `SNOWFALL`


In [36]:
# Use the `mutate()` function to apply min-max normalization on columns 
# `RENTED_BIKE_COUNT`, `TEMPERATURE`, `HUMIDITY`, `WIND_SPEED`, `VISIBILITY`, `DEW_POINT_TEMPERATURE`, `SOLAR_RADIATION`, `RAINFALL`, `SNOWFALL`
dataframe <- dataframe %>% mutate(RENTED_BIKE_COUNT = (RENTED_BIKE_COUNT - min(RENTED_BIKE_COUNT)/(max(RENTED_BIKE_COUNT))-min(RENTED_BIKE_COUNT)),
                                  TEMPERATURE = (TEMPERATURE - min(TEMPERATURE)/(max(TEMPERATURE))-min(TEMPERATURE)),
                                  HUMIDITY = (HUMIDITY - min(HUMIDITY)/(max(HUMIDITY))-min(HUMIDITY)),
                                  WIND_SPEED = (WIND_SPEED - min(WIND_SPEED)/(max(WIND_SPEED))-min(WIND_SPEED)),
                                  VISIBILITY = (VISIBILITY - min(VISIBILITY)/(max(VISIBILITY))-min(VISIBILITY)),
                                  DEW_POINT_TEMPERATURE = (DEW_POINT_TEMPERATURE - min(DEW_POINT_TEMPERATURE)/(max(DEW_POINT_TEMPERATURE))-min(DEW_POINT_TEMPERATURE)),
                                  SOLAR_RADIATION = (SOLAR_RADIATION - min(SOLAR_RADIATION)/(max(SOLAR_RADIATION))-min(SOLAR_RADIATION)),
                                  RAINFALL = (RAINFALL - min(RAINFALL)/(max(RAINFALL))-min(RAINFALL)),
                                  SNOWFALL = (SNOWFALL - min(SNOWFALL)/(max(SNOWFALL))-min(SNOWFALL)))

In [37]:
# Print the summary of the dataset again to make sure the numeric columns range between 0 and 1
summary(dataframe)

     DATE           RENTED_BIKE_COUNT  TEMPERATURE      HUMIDITY    
 Length:8760        Min.   : NA       Min.   : NA    Min.   : 0.00  
 Class :character   1st Qu.: NA       1st Qu.: NA    1st Qu.:42.00  
 Mode  :character   Median : NA       Median : NA    Median :57.00  
                    Mean   :NaN       Mean   :NaN    Mean   :58.23  
                    3rd Qu.: NA       3rd Qu.: NA    3rd Qu.:74.00  
                    Max.   : NA       Max.   : NA    Max.   :98.00  
                    NA's   :8760      NA's   :8760                  
   WIND_SPEED      VISIBILITY        DEW_POINT_TEMPERATURE SOLAR_RADIATION 
 Min.   :0.000   Min.   :  -0.0135   Min.   : 1.125        Min.   :0.0000  
 1st Qu.:0.900   1st Qu.: 912.9865   1st Qu.:27.025        1st Qu.:0.0000  
 Median :1.500   Median :1670.9865   Median :36.825        Median :0.0100  
 Mean   :1.725   Mean   :1409.8123   Mean   :35.799        Mean   :0.5691  
 3rd Qu.:2.300   3rd Qu.:1972.9865   3rd Qu.:46.525        3rd Qu.:0

In [38]:
# Save the dataset as `seoul_bike_sharing_converted_normalized.csv`
write_csv(dataframe, "seoul_bike_sharing_converted_normalized.csv")

## Standardize the column names again for the new datasets


Since you have added many new indicator variables, you need to standardize their column names again by using the following code:


In [39]:
# Dataset list
dataset_list <- c('seoul_bike_sharing.csv', 'seoul_bike_sharing_converted.csv', 'seoul_bike_sharing_converted_normalized.csv')

for (dataset_name in dataset_list){
    # Read dataset
    dataset <- read_csv(dataset_name)
    # Standardized its columns:
    # Convert all columns names to uppercase
    names(dataset) <- toupper(names(dataset))
    # Replace any white space separators by underscore, using str_replace_all function
    names(dataset) <- str_replace_all(names(dataset), " ", "_")
    # Save the dataset back
    write.csv(dataset, dataset_name, row.names=FALSE)
}

Warning message:
“Missing column names filled in: 'X1' [1]”Parsed with column specification:
cols(
  X1 = col_double(),
  DATE = col_character(),
  RENTED_BIKE_COUNT = col_double(),
  HOUR = col_double(),
  TEMPERATURE = col_double(),
  HUMIDITY = col_double(),
  WIND_SPEED = col_double(),
  VISIBILITY = col_double(),
  DEW_POINT_TEMPERATURE = col_double(),
  SOLAR_RADIATION = col_double(),
  RAINFALL = col_double(),
  SNOWFALL = col_double(),
  SEASONS = col_character(),
  HOLIDAY = col_character(),
  FUNCTIONING_DAY = col_character()
)
Parsed with column specification:
cols(
  .default = col_double(),
  DATE = col_character()
)
See spec(...) for full column specifications.
Parsed with column specification:
cols(
  .default = col_double(),
  DATE = col_character(),
  RENTED_BIKE_COUNT = col_logical(),
  TEMPERATURE = col_logical()
)
See spec(...) for full column specifications.


# Next Steps


Great! Now that you have processed all of the necessary datasets, you are ready to perform exploratory data analysis to get some inital insights from them.


## Authors

<a href="https://www.linkedin.com/in/yan-luo-96288783/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkRP0321ENSkillsNetwork25371262-2022-01-01" target="_blank">Yan Luo</a>


### Other Contributors

Jeff Grossman


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description      |
| ----------------- | ------- | ---------- | ----------------------- |
| 2021-04-08        | 1.0     | Yan        | Initial version created |
|                   |         |            |                         |
|                   |         |            |                         |

## <h3 align="center"> © IBM Corporation 2021. All rights reserved. <h3/>
